rust-client/src/main.rs

```diff
@@ -54,233 +55,378 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
-            Multiaddr::empty()
-                .with("0.0.0.0".parse::<Ipv4Addr>().unwrap().into())
-                .with(Protocol::Tcp(0)),
-        )
-        .unwrap();
+    swarm.listen_on(
+        Multiaddr::empty()
+            .with("0.0.0.0".parse::<Ipv4Addr>()?.into())
+            .with(Protocol::Tcp(0)),
+    )?;
 
     // Wait to listen on all interfaces.
-    block_on(async {
-        let mut delay = futures_timer::Delay::new(std::time::Duration::from_secs(1)).fuse();
-        loop {
-            futures::select! {
-                event = swarm.next() => {
-                    match event.unwrap() {
-                        SwarmEvent::NewListenAddr { address, .. } => {
-                            info!("Listening on {:?}", address);
-                        }
-                        event => panic!("{:?}", event),
+    let mut delay = futures_timer::Delay::new(std::time::Duration::from_secs(1)).fuse();
+    loop {
+        futures::select! {
+            event = swarm.select_next_some() => {
+                match event {
+                    SwarmEvent::NewListenAddr { address, .. } => {
+                        info!("Listening on {:?}", address);
                     }
+                    event => panic!("{:?}", event),
                 }
-                _ = delay => {
-                    // Likely listening on all interfaces now, thus continuing by breaking the loop.
-                    break;
-                }
+            }
+            _ = delay => {
+                // Likely listening on all interfaces now, thus continuing by breaking the loop.
+                break;
             }
         }
-    });
-
-    let request = tonic::Request::new(grpc::RegisterRequest {
-        client_id: local_peer_id.to_bytes(),
-        agent_version: agent_version(),
-        // TODO: Fill.
-        protocols: vec![],
-    });
-
-    let _response = client.register(request).await?;
-
-    let request = tonic::Request::new(grpc::GetAddrInfoRequest {
-        client_id: local_peer_id.to_bytes(),
-    });
-
-    let response = client.get_addr_info(request).await?.into_inner();
-
-    let remote_peer_id = PeerId::from_bytes(&response.remote_id)?;
-    let remote_addrs = response
-        .multi_addresses
-        .into_iter()
-        .map(|a| Multiaddr::try_from(a))
-        .collect::<Result<Vec<_>, libp2p::multiaddr::Error>>()?;
-
-    info!(
-        "Attempting to punch through to {:?} via {:?}.",
-        remote_peer_id, remote_addrs
-    );
-
-    let start_time = SystemTime::now();
-
-    let result = if remote_addrs
-        .iter()
-        .all(|a| a.iter().any(|p| p == libp2p::multiaddr::Protocol::Quic))
-    {
-        Err(())
-    } else {
+    }
+
+    for _ in 0..ROUNDS {
+        let request = tonic::Request::new(grpc::RegisterRequest {
+            client_id: local_peer_id.to_bytes(),
+            agent_version: agent_version(),
+            // TODO: Automatically fill protocol list instead of hard coding it.
```

Comment:
As a hack, you could do something along the lines of:

https://github.com/libp2p/rust-libp2p/blob/7a1147877a52c7150190acdef35676cdac370299/swarm/src/lib.rs#L1326-L1333

Reply:
Thanks, done in a0cdee1.